bat.d/41-framework

```diff
@@ -0,0 +1,536 @@
+    if [ -n "$X_BAT_PLUGIN_SIMULATE" ]; then
+        if [ "$X_BAT_PLUGIN_SIMULATE" = "$_batdrv_plugin" ]; then
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.simulate"
+        else
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.simulate_skip"
+            return 1
+        fi
+    elif wordinlist "$_batdrv_plugin" "$X_BAT_PLUGIN_DENYLIST"; then
+        echo_debug "bat" "batdrv_init.${_batdrv_plugin}.denylist"
+        return 1
+    else
+        # check if hardware matches
+        if ! batdrv_is_framework; then
+            echo_debug "bat" "batdrv_init.${_batdrv_plugin}.no_match"
+            return 1
+        fi
+    fi
+
+    # presume no features at all
+    _natacpi=128
+    # shellcheck disable=SC2034
+    _bm_read="natacpi"
+    _bm_thresh="none"
+    # shellcheck disable=SC2034
+    _bm_dischg="none"
+    _bn_stop=""
+    _batteries=""
+	# framework doesn't support charge start threshold
+    _bt_def_stop=60
+
+    # iterate batteries and check for native kernel ACPI
+    local bd bs
+    local done=0
+    for bd in "$ACPIBATDIR"/BAT[01]; do
+        if [ "$(read_sysf "$bd/present")" = "1" ]; then
+            # record detected batteries and directories
+            bs=${bd##/*/}
+            if [ -n "$_batteries" ]; then
+                _batteries="$_batteries $bs"
+            else
+                _batteries="$bs"
+            fi
+            # skip natacpi detection for 2nd and subsequent batteries
+            [ $done -eq 1 ] && continue
+
+            done=1
+            if [ "$NATACPI_ENABLE" = "0" ]; then
+                # natacpi disabled in configuration --> skip actual detection
+                _natacpi=32
+                continue
+            fi
+
+            if [ -f "$bd/charge_control_end_threshold" ]; then
+                # threshold sysfiles exist
+                _bn_stop="charge_control_end_threshold"
+                _natacpi=254
+            else
+                # nothing detected
+                _natacpi=254
+                continue
+            fi
+
+            if readable_sysf "$bd/$_bn_stop"; then
+                # threshold sysfiles are actually readable
+                _natacpi=0
+                _bm_thresh="natacpi"
+            fi
+        fi
+    done
+
+    # quit if no battery detected, there is no point in activating the plugin
+    if [ -z "$_batteries" ]; then
+        echo_debug "bat" "batdrv_init.${_batdrv_plugin}.no_batteries"
+        return 2
+    fi
+
+    # shellcheck disable=SC2034
+    _batdrv_selected=$_batdrv_plugin
+    echo_debug "bat" "batdrv_init.${_batdrv_plugin}: batteries=$_batteries; natacpi=$_natacpi; thresh=$_bm_thresh; bn_stop=$_bn_stop;"
+    return 0
+}
+
+batdrv_select_battery () {
+    # determine battery acpidir and sysfiles
+    # $1: BAT0/BAT1/DEF
+    # # rc: 0=bat exists/1=bat non-existent
+    # retval: $_bat_str:   BAT0/BAT1/BATT;
+    #         $_bd_read:   directory with battery data sysfiles;
```

Comment:
unfortunately, I had to make changes to the return values in the meantime, you have to additionally return `_bt_cfg_bat`, see here: https://github.com/linrunner/TLP/blob/main/bat.d/05-thinkpad#L352